lib/vagrant/bundler.rb

```diff
@@ -62,57 +195,113 @@ def initialize
+      composed_set = generate_vagrant_set
+
+      if solution_file&.valid?
+        @logger.debug("loading cached solution set")
+        solution = solution_file.dependency_list.map do |dep|
+          spec = composed_set.find_all(dep).first
+          if !spec
+            @logger.warn("failed to locate specification for dependency - #{dep}")
+            @logger.warn("invalidating solution file - #{solution_file}")
+            solution_file.invalidate!
+            break
+          end
+          dep_r = Gem::Resolver::DependencyRequest.new(dep, nil)
+          Gem::Resolver::ActivationRequest.new(spec, dep_r)
+        end
       end
 
-      @logger.debug("Current generated plugin dependency list: #{plugin_deps}")
+      if !solution_file&.valid?
+        @logger.debug("generating solution set for configured plugins")
+        # Add HashiCorp RubyGems source
+        if !Gem.sources.include?(HASHICORP_GEMSTORE)
+          sources = [HASHICORP_GEMSTORE] + Gem.sources.sources
+          Gem.sources.replace(sources)
+        end
 
-      # Load dependencies into a request set for resolution
-      request_set = Gem::RequestSet.new(*plugin_deps)
-      # Never allow dependencies to be remotely satisfied during init
-      request_set.remote = false
+        # Generate dependencies for all registered plugins
+        plugin_deps = plugins.map do |name, info|
+          Gem::Dependency.new(name, info['installed_gem_version'].to_s.empty? ? '> 0' : info['installed_gem_version'])
+        end
 
-      repair_result = nil
-      begin
-        # Compose set for resolution
-        composed_set = generate_vagrant_set
-        # Resolve the request set to ensure proper activation order
-        solution = request_set.resolve(composed_set)
-      rescue Gem::UnsatisfiableDependencyError => failure
-        if repair
-          raise failure if @init_retried
-          @logger.debug("Resolution failed but attempting to repair. Failure: #{failure}")
-          install(plugins)
-          @init_retried = true
-          retry
-        else
-          raise
+        @logger.debug("Current generated plugin dependency list: #{plugin_deps}")
+
+        # Load dependencies into a request set for resolution
+        request_set = Gem::RequestSet.new(*plugin_deps)
+        # Never allow dependencies to be remotely satisfied during init
+        request_set.remote = false
+
+        repair_result = nil
+        begin
+          @logger.debug("resolving solution from available specification set")
+          # Resolve the request set to ensure proper activation order
+          solution = request_set.resolve(composed_set)
+          @logger.debug("solution set for configured plugins has been resolved")
+        rescue Gem::UnsatisfiableDependencyError => failure
+          if repair
+            raise failure if @init_retried
+            @logger.debug("Resolution failed but attempting to repair. Failure: #{failure}")
+            install(plugins)
```

Comment:
I seem to get a stacktrace when trying to install a plugin:

```
brian@localghost:vagrant-sandbox % be vagrant plugin install vagrant-hostmanager                         ±[●][master]
Installing the 'vagrant-hostmanager' plugin. This can take a few minutes...
Fetching vagrant-hostmanager-1.8.9.gem
Traceback (most recent call last):
        28: from /home/brian/.rvm/gems/ruby-2.6.5@vagrant-sandbox/bin/ruby_executable_hooks:24:in `<main>'
        27: from /home/brian/.rvm/gems/ruby-2.6.5@vagrant-sandbox/bin/ruby_executable_hooks:24:in `eval'
        26: from /home/brian/.rvm/gems/ruby-2.6.5@vagrant-sandbox/bin/vagrant:23:in `<main>'
        25: from /home/brian/.rvm/gems/ruby-2.6.5@vagrant-sandbox/bin/vagrant:23:in `load'
        24: from /home/brian/go/src/github.com/hashicorp/vagrant/bin/vagrant:182:in `<top (required)>'
        23: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/environment.rb:290:in `cli'
        22: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/cli.rb:66:in `execute'
        21: from /home/brian/go/src/github.com/hashicorp/vagrant/plugins/commands/plugin/command/root.rb:66:in `execute'
        20: from /home/brian/go/src/github.com/hashicorp/vagrant/plugins/commands/plugin/command/install.rb:69:in `execute'
        19: from /home/brian/go/src/github.com/hashicorp/vagrant/plugins/commands/plugin/command/install.rb:69:in `each'
        18: from /home/brian/go/src/github.com/hashicorp/vagrant/plugins/commands/plugin/command/install.rb:70:in `block in execute'
        17: from /home/brian/go/src/github.com/hashicorp/vagrant/plugins/commands/plugin/command/base.rb:14:in `action'
        16: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/action/runner.rb:102:in `run'
        15: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/util/busy.rb:19:in `busy'
        14: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/action/runner.rb:102:in `block in run'
        13: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/action/builder.rb:116:in `call'
        12: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/action/warden.rb:57:in `call'
        11: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
        10: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/action/warden.rb:57:in `call'
         9: from /home/brian/go/src/github.com/hashicorp/vagrant/plugins/commands/plugin/action/install_gem.rb:30:in `call'
         8: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/plugin/manager.rb:148:in `install_plugin'
         7: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/plugin/manager.rb:138:in `block in install_plugin'
         6: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/bundler.rb:332:in `install'
         5: from /home/brian/go/src/github.com/hashicorp/vagrant/lib/vagrant/bundler.rb:577:in `internal_install'
         4: from /home/brian/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/request_set.rb:298:in `install_into'
         3: from /home/brian/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/request_set.rb:322:in `install_hooks'
         2: from /home/brian/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/request_set.rb:322:in `each'
         1: from /home/brian/.rvm/rubies/ruby-2.6.5/lib/ruby/site_ruby/2.6.0/rubygems/request_set.rb:323:in `block in install_hooks'
/home/brian/.rvm/rubies/ruby-2.6.5/lib/ruby/2.6.0/rdoc/rubygems_hook.rb:52:in `generation_hook': undefined method `include?' for false:FalseClass (NoMethodError)
```

For what it's worth it doesn't seem to happen when I'm not running Vagrant from source (i.e. putting this branch inside a Vagrant package). But also I don't get this exception if I try to install a plugin on the master branch from source.